### PR TITLE
JIT: Optimize movzx after setcc

### PR DESCRIPTION
**Pull request by EgorBo**
Clear target reg via `xor reg, reg` instead of relying on zero-extend after SETCC which is heavier and slower
```csharp
bool Test(int x) => x == 42;
```
```diff
; Method Test(int):bool:this
G_M55728_IG01:              ;; offset=0000H
G_M55728_IG02:              ;; offset=0000H
+      33C0                 xor      eax, eax
       83FA2A               cmp      edx, 42
       0F94C0               sete     al
-      0FB6C0               movzx    rax, al
G_M55728_IG03:              ;; offset=0009H
       C3                   ret      
-; Total bytes of code: 10
+; Total bytes of code: 9
```
Nice diffs: 
```
benchmarks.run.Linux.x64.checked.mch:
Total bytes of delta: -5230 (-0.03 % of base)

coreclr_tests.pmi.Linux.x64.checked.mch:
Total bytes of delta: -210860 (-0.16 % of base)

libraries.crossgen2.Linux.x64.checked.mch:
Total bytes of delta: -4420 (-0.06 % of base)

libraries.pmi.Linux.x64.checked.mch:
Total bytes of delta: -25312 (-0.05 % of base)

libraries_tests.pmi.Linux.x64.checked.mch:
Total bytes of delta: -49314 (-0.04 % of base)
```

### CODE REVIEW

**Comment by EgorBo** [comment]
it's not safe apparently, xor clears flags

**Comment by jakobbotsch** [comment]
> it's not safe apparently, xor clears flags

In which cases did it matter? Doesn't all of these cases emit a `cmp` after?

**Comment by EgorBo** [comment]
> > it's not safe apparently, xor clears flags
> 
> In which cases did it matter? Doesn't all of these cases emit a `cmp` after?

I wasn't able to detect - tests were keeping failing, I suspect some complex logic in emitInsBinary?

**Comment by EgorBo** [comment]
Ah, I think I've found out why

**Comment by EgorBo** [comment]
My understanding that it used to crash on things like
```
xor      eax, eax
cmp      dword ptr [rax+8], 0
```

**Comment by EgorBo** [comment]
cc @dotnet/jit-contrib @jakobbotsch

**Review by jakobbotsch — APPROVED** [review]
LGTM. What do the diffs look like if you do something less conservative? e.g.
```cpp
regMaskTP GenTree::gtGetContainedRegMask()
{
    regMaskTP mask = gtGetRegMask();
    for (GenTree* operand : Operands())
    {
        if (operand->isContained())
        {
            mask |= operand->gtGetContainedRegMask();
        }
    }

    return mask;
}
```
and do the optimization unless `(op1->gtGetContainedRegMask() | op2->gtGetContainedRegMask()) & genRegMask(targetReg)`? (not sure if that's completely right)

**Comment by kasperk81** [comment]
> Nice diffs:

nice. and regressions are just cse diffs or something more interesting?

File: src/coreclr/jit/gentree.cpp [hunk 1]
```diff
@@ -862,6 +862,26 @@ unsigned GenTree::GetMultiRegCount(Compiler* comp) const
...
+//
+regMaskTP GenTree::gtGetContainedRegMask()
+{
+    regMaskTP mask = GetRegNum() != REG_NA ? gtGetRegMask() : 0;
+    for (GenTree* operand : Operands())
+    {
+        mask |= operand->gtGetContainedRegMask();
+    }
+    return mask;
+}
```
[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/jit/gentree.cpp:883]: I.e. I guess this is a more correct version than my original:
```cpp
regMaskTP GenTree::gtGetContainedRegMask()
{
    if (!isContained())
        return gtGetRegMask();

    regMaskTP mask = 0;
    for (GenTree* operand : Operands())
    {
        mask |= operand->gtGetContainedRegMask();
    }
    return mask;
}
```

**Comment by kasperk81** [comment]
thanks. overall it is nice win (regression in bignumber format method is a red-herring).
c++ compiler generated interesting asm with that repro (almost half the no. of lines)

```
IsResultCode(ResultCode):
        cmp     edi, 34
        ja      .L2
        movabs  rdx, -30068932608
        mov     eax, 1
        bt      rdx, rdi
        jc      .L2
        ret
.L2:
        cmp     edi, 36
        sete    al
        cmp     edi, 48
        sete    dl
        or      eax, edx
        ret
```

**Comment by EgorBo** [comment]
> c++ compiler generated interesting asm with that repr

Yeah, the bitmask hack is a nice thing to have, e.g. here it's used in C# https://github.com/dotnet/runtime/blob/main/src%2Flibraries%2FSystem.Private.CoreLib%2Fsrc%2FSystem%2FRuntime%2FCompilerServices%2FRuntimeHelpers.cs#L87-L89 (because jit is not yet able to optimize it)

**Comment by EgorBo** [comment]
@jakobbotsch thank you, diffs are even bigger now

**Pull request merged**